llvm-general/src/LLVM/General/Internal/FFI/TargetC.cpp

```diff
@@ -338,22 +316,42 @@ void LLVM_General_InitializeAllTargets() {
   // None of the other components are bound yet
 }
 
+// This is identical to LLVMTargetMachineEmit but LLVM doesn’t expose this function so we copy it here.
 LLVMBool LLVM_General_TargetMachineEmit(
-	LLVMTargetMachineRef TM,
-	LLVMModuleRef M,
-	LLVMCodeGenFileType codeGenFileType,
-	char **ErrorMessage,
-	raw_pwrite_stream &dest
+    LLVMTargetMachineRef T,
+    LLVMModuleRef M,
+    raw_pwrite_stream &OS,
+    LLVMCodeGenFileType codegen,
+    char **ErrorMessage
 ) {
-	TargetMachine &tm = *unwrap(TM);
-    legacy::PassManager passManager;
-	if (tm.addPassesToEmitFile(passManager, dest, unwrap(codeGenFileType)))
-{
-		*ErrorMessage = strdup("TargetMachine can't emit a file of this type");
-		return true;
-	}
+  TargetMachine* TM = unwrap(T);
+  Module* Mod = unwrap(M);
+
+  legacy::PassManager pass;
 
-	passManager.run(*unwrap(M));
-	return false;
+  std::string error;
+
+  Mod->setDataLayout(TM->createDataLayout());
+
+  TargetMachine::CodeGenFileType ft;
+  switch (codegen) {
+    case LLVMAssemblyFile:
+      ft = TargetMachine::CGFT_AssemblyFile;
+      break;
+    default:
+      ft = TargetMachine::CGFT_ObjectFile;
+      break;
+  }
+  if (TM->addPassesToEmitFile(pass, OS, ft)) {
+    error = "TargetMachine can't emit a file of this type";
+    *ErrorMessage = strdup(error.c_str());
```

Comment:
This implementation is exactly identical to the one in LLVM (as I mentioned in the comment). I don’t see any good reason to diverge from this as it will only mean more effort when updating and we have to figure out why something broke.